### DISCUSSION TITLE
Incomplete topologies and unaccounted external momenta

### DISCUSSION THREAD

**Discussion by ExAstrisScientia** [FeynCalc/feyncalc#249] [Ideas]
Hi, 

we have a question and potential feature request for FCLoopBasisFindCompletion.
Our problem is the following: In a two-loop calculation involving two external momenta we find a topology which only depends on one of the external momenta, while the numerators involve both. This causes FCLoopApplyTopologyMappings to leave scalar products of loop momenta with the unused external momentum alone, and the proceeding calculation fails. 

A simple workaround we found (and at least to us novices in the art of multiloop calculations this seems to not screw anything up) is to manually define new topologies by adding the missing scalar products as new propagators to the old topologies, in the same way FCLoopBasisFindCompletion seems to do for cases where an individual scalar product cannot be resolved in an incomplete basis: by adding an eikonal looking propagator with the missing scalar product explicitly. The topologies appearing in the subsequent calculation then have a longer list of arguments (i.e. propagator powers), where the new entries are mainly populated by 0s and -1s. Tensor reduction and interfacing with FIRE seem to work and we don't encounter any problems. 

With this in mind: Is there an official function, or flag, or option to set, which does what we did already, maybe in a more canonical way, and we simply didn't see it? Or is what we did actually a Really Bad Idea™, and should not be done for reasons we simply aren't experienced enough to understand?

If not, and if this is actually an issue that isn't addressed in the code yet, and if the developers think this might be a good addition, we'd like to request adding a feature to FCLoopBasisFindCompletion to pad out the topology with the missing scalar products:

Currently, the output of FCLoopFindTopologies returns a list of the external momenta it encountered in the fourth argument of the topology expression. This list seems to only contain the external momenta it finds in the set of propagators, i.e. this is a subset of the set of all external momenta (In our case the output only contains {p2}). We could envision an option to FCLoopBasisFindCompletion specifying the actual set of external momenta (in our case "{p1,p2}"), and FCLoopBasisFindCompletion subsequently adds the scalar products between all of the loop momenta and the unaccounted momenta via new propagators. So something like this:

OldTopology:
![Screenshot from 2024-02-15 15-34-25](https://github.com/FeynCalc/feyncalc/assets/73316366/250ebc1c-28fc-4eb1-ba98-59927a665d52)

Scalar products that are missing: 

> Pair[Momentum[k1],Momentum[p1]]
> Pair[Momentum[k2],Momentum[p1]]

When calling

> FCLoopBasisFindCompletion[OldTopology,ExplicitExternal->{p1,p2}]

the output should be something like:

![Screenshot from 2024-02-15 15-35-29](https://github.com/FeynCalc/feyncalc/assets/73316366/1a61644b-fa7e-43f7-8cc8-14efe110e58d)

We're not quite sure whether the kinematic replacements are as easy to implement, but already the completion of the basis would solve most of the problem.

(If needed we can send a MWE Mathematica file, to highlight the problem in its original setting).

Looking forward to hearing from you,
All the best,
Rudi and Giuseppe

**Comment by vsht** [comment]
Hi,

it's nice to see someone trying to actually make use of the new multiloop capabilities, so my feedback would be as follows:

Personally, in the case of

> Our problem is the following: In a two-loop calculation involving two external momenta we find a topology which only depends on one of the external momenta, while the numerators involve both. This causes FCLoopApplyTopologyMappings to leave scalar products of loop momenta with the unused external momentum alone, and the proceeding calculation fails.

I would just do a tensor reduction using `FCLoopTensorReduce`. This would decouple the external momentum that doesn't appear in the propagators and immediately solve the issue.

```
topo = FCTopology[
  fctopology1, {SFAD[{k2, mt^2}], SFAD[k1 - k2], SFAD[{k1, mt^2}], 
   SFAD[k1 - k2 + p2], SFAD[{k1 + p2, mt}]}, {k1, 
   k2}, {p2}, {Hold[SPD][p2, p2] -> 0, Hold[SPD][p1, p2] -> p12}, {}]

amp = gliProduct[SPD[k1, p2] SPD[k2, p1], GLI[fctopology1, {1, 1, 1, 1, 1}]]

FCLoopTensorReduce[amp, {topo}, Head -> gliProduct]
(* gliProduct[(SPD[k1, p2] SPD[k2, p2] SPD[p1, p2])/SPD[p2, p2], GLI[fctopology1, {1, 1, 1, 1, 1}]]*)
```

> With this in mind: Is there an official function, or flag, or option to set, which does what we did already, maybe in a more canonical way, and we simply didn't see it? Or is what we did actually a Really Bad Idea™, and should not be done for reasons we simply aren't experienced enough to understand?

May be I'm missing something here, but how did you manage to finish the calculation with `fctopology1c`? While `fctopology1` has a complete basis, your `fctopology1c` is overcomplete.
```
topo2 = FCTopology[
  fctopology1C, {SFAD[{k2, mt^2}], SFAD[k1 - k2], SFAD[{k1, mt^2}], SFAD[k1 - k2 + p2], SFAD[{k1 + p2, mt}],
   SFAD[{{0, k1 . p1}}], SFAD[{{0, k1 . p2}}]}, {k1, k2}, {p1, p2}, {Hold[SPD][p2, p2] -> 0, Hold[SPD][p1, p2] -> p12}, {}]
FCLoopBasisOverdeterminedQ[topo2]
(*True*)
```

Apart from that, for what you have in mind, `FCLoopBasisFindCompletion` has an option `Method` that allows you to specify a list of denominators to choose from when completing the basis

```
topoTest =  FCTopology[  fctopologyX, {SFAD[q1 + p, q2 - k] SPD[q1, q2]}, {q1, q2}, {p,    k}, {}, {}]
FCLoopBasisFindCompletion[topoTest,  Method -> {SFAD[{q2 + k, m}], SFAD[{q1 - p, m}], SPD[p, q2],    SPD[k, q1]}, FCE -> True]
```
Cheers,
Vlad

**Comment by ExAstrisScientia** [comment]
Hi Vlad,

apologies for the delay, we were battling some FeynArts problems along the way.

You are right that we can use the tensor reduction, but there is a caveat: If we set external kinematics early then this can fail. In our case specifically if we set

>SPD[p2,p2]=0

the tensor reduction doesn't do anything. (This is also visible in the code snipped you posted, the output is only well defined if p2^2 doesn't vanish.) We had set this, and then the tensor reduction simply doesn't resolve some of the appearing scalar products. So if we keep the scalar products around until the end, this does indeed seem to solve the problem.

However, it is apparently also possible to extend the topologies in the manner we described, in such a way that the new topologies remain not overdetermined. The case we listed is one such case, you made a small typo in your code snippet: The last propagator should be k2.p1, not k1.p2. If you use k2.p1, FCLoopBasisOverdeterminedQ confirms that the new topology is not overdetermined.

When using the extended topology, the tensor reduction also goes through if we set the kinematics, and we get a final result after FIRE which is written in terms of a different set of master integrals.
We'll now probably have a look at whether the two FIRE outputs are indeed the same, because it would of course be very nice if we could use kinematics during the calculation, rather than only at the very end.

In any case, your suggestion does indeed solve our problem, though we may approach you again if it turns out that the extension reduces the complexity of intermediate steps :)

Cheers,
Rudi and Giuseppe

**Reply by vsht** [reply]
Hi,

nice to hear that it helped. Tensor reduction is indeed not possible for zero Gram determinants, but extending the topology is probably 
also not something that is going to work in all cases.

Anyhow, thanks for your feedback when using the new multiloop toolset.

Cheers,
Vlad
